File: questao8.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <locale.h>
#include <stdbool.h>
#include <time.h> 

typedef struct Jogador{

    int id;
    char nome[100];
    int altura;
    int peso;
    char universidade[100];
    int anoNascimento;
    char cidadeNascimento[100];
    char estadoNascimento[100];

}Jogador;

//--------- chamada dos metodos --------------
void LerJogador(char entradaID[]);
char* replace(char * s);
void TratarString(char entrada[]);
void insercaoPorCor(int n, int cor, int h);
void shellsort(int n);
void sort();
void Mostrar();
//-----------------------
void id(char id[]);
void Nome(char nome[]);
void Altura(char altura[]);
void Peso(char peso[]);
void Universidade(char universidade[]);
void AnoNascimento(char anoNascimento[]);
void CidadeNascimento(char cidadeNascimento[]);
void EstadoNascimento(char estadoNascimento[]);

// main

Jogador lista[1000];
int contadorjog = 0;
int contador = 0;

int main(){
    clock_t t; 
    t = clock();
    
    char Ids[1000];
    scanf("%s",Ids);

    while(strcmp(Ids,"FIM") != 0 ){             
        LerJogador(Ids);
        contadorjog++;
        scanf("%s",Ids);
    }
    shellsort(contadorjog);
    sort();
    Mostrar();

    // arquivo de matricula========================================================
    t = clock() - t; 
    
    FILE *arq;
    arq = fopen("matrícula_shellsort.txt", "a");
    fprintf(arq, "695161 \t %ld \t %d", t , contador);
    fclose(arq);
    //=============================================================================
    return 0;
}

// """SETS""" 

void id(char id[]){
    lista[contadorjog].id = atoi(id);
}
void Nome(char nome[]){
    strcpy(lista[contadorjog].nome,nome);
}
void Altura(char altura[]){
    lista[contadorjog].altura = atoi(altura);
}
void Peso(char peso[]){
    lista[contadorjog].peso = atoi(peso);
}
void AnoNascimento(char anoNascimento[]){
    lista[contadorjog].anoNascimento = atoi(anoNascimento);
}
void Universidade(char universidade[]){
    strcpy(lista[contadorjog].universidade,universidade);
}
void CidadeNascimento(char cidadeNascimento[]){
    strcpy(lista[contadorjog].cidadeNascimento,cidadeNascimento);
}
void EstadoNascimento(char estadoNascimento[]){
    strcpy(lista[contadorjog].estadoNascimento,estadoNascimento);
}

// Shell sort 

//=============================================================================
void insercaoPorCor(int n, int cor, int h){
    for (int i = (h + cor); i < n; i+=h) {
        Jogador tmp = lista[i];
        int j = i - h;
        while ((j >= 0) && ((lista[j].peso > tmp.peso || (lista[j].peso == tmp.peso && strcmp(lista[j].nome, tmp.nome)>0 )))) {
            lista[j + h] = lista[j];
            j-=h;
            contador++;
        }
        lista[j + h] = tmp;
    }
}

void shellsort(int n) {
    Jogador tmp; 
   
    for(int i=0; i<n; i++){
        for(int j=i+1; j<n; j++){
            if(lista[i].altura == lista[j].altura){
                if( strcmp(lista[i].nome,lista[j].nome)>0){
                   tmp=lista[i];
                   lista[i]=lista[j];
                   lista[j]=tmp;
                }
            }else{
                j=n;     
            }
        }
    }
}

void sort(){
    int h = 1;
   
    do{
        h = (h * 3) + 1;
    }while (h < contadorjog);

    do{
        h /= 3;
        for (int cor = 0; cor < h; cor++){
            insercaoPorCor(contadorjog, cor, h);
            contador++;
        }
    } while (h != 1);
}
//=============================================================================
// LEITURA

void LerJogador(char entradaID[]){

    char entradas[1000];
    char *stringsep;
    char *virgula;
    FILE *caminho = fopen("/tmp/players.csv","r");
    // C:\\Users\\WazX\\Desktop\\aeds2-master\\tps\\entrada e saida\\players.csv && /tmp/players.csv

    do{
        fgets(entradas,1000,caminho);
        strcpy(entradas , replace(entradas));
        stringsep = strdup(entradas);
        virgula = strtok(stringsep,",");
        if (strcmp(virgula,entradaID) == 0 ){
            TratarString(entradas);
        }  
    }while(!feof(caminho) && strcmp(virgula,entradaID) != 0);
    fclose(caminho);

}

void TratarString(char entrada[]){

    //ID
    strcpy(entrada, strtok(entrada, ",")); 
    id(entrada);    
    //Nome
    strcpy(entrada, strtok(NULL, ","));
    Nome(entrada);
    //ALTURA
    strcpy(entrada, strtok(NULL, ","));
    Altura(entrada);  
    //PESO
    strcpy(entrada, strtok(NULL, ","));
    Peso(entrada);
    //UNIVERSIDADE
    strcpy(entrada, strtok(NULL, ","));
    Universidade(entrada);   
    //ANO DE NASCIMENTO
    strcpy(entrada, strtok(NULL, ","));
    AnoNascimento(entrada);
    //CIDADE DE NASCIMENTO
    strcpy(entrada, strtok(NULL, ","));
    CidadeNascimento(entrada);
    //ESTADO DE NASCIMENTO
    strcpy(entrada, strtok(NULL, ","));
    EstadoNascimento(entrada);
}

// separador de virgulas

char* replace(char s[]){

    char novaentrada[400]="";
    int cont = 0;
    int ult = strlen(s);

    for (int i = 0; i < ult-1; i++){
        if ( s[i] == ',' && s[i-1] == ','){
            strcat(novaentrada,"nao informado");
            cont = strlen(novaentrada);
        }
        novaentrada[cont] = s[i];
        cont++;
    }

    if(novaentrada[strlen(novaentrada)-1] == ','){
        strcat(novaentrada,"nao informado");
    }
    
    char *resp = novaentrada;
    return resp;
}

// CLONE

/*Jogador JogadorClone(const Jogador* const self){
    Jogador clone;

    clone.id = self->nome;
    clone.nome = self->nome;
    clone.altura = self->altura;
    clone.peso = self->peso;
    clone.universidade = self->universidade;
    clone.anoNascimento = self->anoNascimento;
    clone.cidadeNascimento = self->cidadeNascimento;
    clone.estadoNascimento = self->estadoNascimento;
}*/

// MOSTRAR

void Mostrar(){

    for (int i = 0; i < contadorjog; i++){
        printf("[%d", lista[i].id);
        printf("%s"," ## ");
        printf("%s", lista[i].nome);
        printf("%s"," ## ");
        printf("%d", lista[i].altura);
        printf("%s"," ## ");
        printf("%d", lista[i].peso);
        printf("%s"," ## ");
        printf("%d", lista[i].anoNascimento);
        printf("%s"," ## ");
        printf("%s", lista[i].universidade);
        printf("%s"," ## ");
        printf("%s", lista[i].cidadeNascimento);
        printf("%s"," ## ");
        printf("%s", lista[i].estadoNascimento);
        printf("]\n");   

    }
}

File: TP4/questao8.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;

class Jogador{

    public static Jogador[] jogad = new Jogador[1000];
    public static int tamJog = 0;

    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;

    //CONSTRUTORES

    public Jogador(){
        this.id = 0;
        this.nome= "";
        this.altura= 0;
        this.peso= 0;
        this.universidade= "";
        this.anoNascimento= 0;
        this.cidadeNascimento= "";
        this.estadoNascimento= "";
    }

    public Jogador(int id,String nome,int altura,int peso,
    String universidade,int anoNascimento,String cidadeNascimento,String estadoNascimento){

        this.id = id;
        this.nome= nome;
        this.altura= altura;
        this.peso= peso;
        this.universidade= universidade;
        this.anoNascimento= anoNascimento;
        this.cidadeNascimento= cidadeNascimento;
        this.estadoNascimento= estadoNascimento;
    }

    //GETS

    public int getId() {
        return id;
    }
    public String getNome() {
        return nome;
    }
    public int getAltura() {
        return altura;
    }public int getPeso() {
        return peso;
    }
    public String getUniversidade() {
        return universidade;
    }
    public int getAnoNascimento() {
        return anoNascimento;
    }
    public String getCidadeNascimento() {
        return cidadeNascimento;
    }
    public String getEstadoNascimento() {
        return estadoNascimento;
    }

    //SETS

    public void setId(int id) {
        this.id = id;
    }
    public void setNome(String nome) {
        this.nome = nome;
    }
    public void setAltura(int altura) {
        this.altura = altura;
    }
    public void setPeso(int peso) {
        this.peso = peso;
    }
    public void setUniversidade(String universidade) {
        this.universidade = universidade;
    }
    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }
    public void setCidadeNascimento(String cidadeNascimento) {
        this.cidadeNascimento = cidadeNascimento;
    }
    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    //CLONE

    public Jogador clone(){

        Jogador clone = new Jogador();

        clone.id = id;
        clone.nome = nome;
        clone.altura = altura;
        clone.peso = peso;
        clone.universidade = universidade;
        clone.anoNascimento = anoNascimento;
        clone.cidadeNascimento = cidadeNascimento;
        clone.estadoNascimento = estadoNascimento;

        return clone;
    }

    //MOSTRAR

    public void mostrar(){
        System.out.println("["+ getId()+" ## "+ getNome()+" ## "+ getAltura()+" ## "+ getPeso()
            +" ## "+ getAnoNascimento()+" ## "+ getUniversidade()+" ## "+getCidadeNascimento()+
            " ## " +getEstadoNascimento()+" ## "+"]");
    }
}

class HashComRehash {
	Jogador tabela[];
	int m;
	final int NULO = -1;

	public HashComRehash() {
		this(25);
	}

	public HashComRehash(int m) {
		this.m = m;
		this.tabela = new Jogador[this.m];
		for (int i = 0; i < m; i++) {
			tabela[i] = null;
		}
	}

	public int h(int elemento) {
		return elemento % m;
	}

	public int reh(int elemento) {
		return (elemento+1)% m;
	}

	public boolean inserir(Jogador elemento) {
		boolean resp = false;
		if (elemento != null) {
			int pos = h(elemento.getAltura());
			if (tabela[pos] == null) {
				tabela[pos] = elemento;
				resp = true;
			} else {
				pos = reh(elemento.getAltura());
				if (tabela[pos] == null) {
				tabela[pos] = elemento;
				resp = true;
				}
			}
		}
		return resp;
	}

    public boolean pesquisar(int altura) {
        boolean resp = false;
        int pos = h(altura);

        // Procura na posição calculada pela função de hash
        if (tabela[pos] != null && tabela[pos].getAltura() == altura) {
            resp = true;
        } else {
            // Procura na posição recalculada pela função de rehash
            pos = reh(altura);
            if (tabela[pos] != null && tabela[pos].getAltura() == altura) {
                resp = true;
            }
        }
        return resp;
    }

	public boolean pesquisar(int elemento , String nomejog) {
        boolean resp = false;
        int pos = h(elemento);
        if (tabela[pos].getNome().compareTo(nomejog) == 0) {
            resp = true;
        }else if (tabela[pos] != null) {
            pos = reh(elemento);
            if (tabela[pos].getNome().compareTo(nomejog) == 0) {
                resp = true;
            }
        }
        return resp;
   }
}

public class questao8{
    public static Jogador jogad[] = new Jogador[1000];
    public static int tamJog = 0;
    public static int contador = 0;
    
    //LER
    public static String ler(String entradaid) throws Exception {
        String entrada = "";
        BufferedReader arq = new BufferedReader(new InputStreamReader(new FileInputStream("/tmp/players.csv")));
        entrada = arq.readLine();
        while(entrada != null){
            String[] Ids = entrada.split(",");
            if(Ids[0].equals(entradaid) == true){
                return entrada;
            }
            entrada = arq.readLine();
        }
        String saida = "";
        
        return saida;
    }

    public static void TratarString(String entrada)throws Exception{
        int tam = entrada.length();
        if(entrada.charAt(tam-1) == ','){
            entrada = entrada + "nao informado";
        }
        entrada = entrada.replaceAll(",,", ",nao informado,");
        String[] jogadores = entrada.split(",");

        
        // ID do jogador
        int id = Integer.parseInt(jogadores[0]);
        jogad[tamJog].setId(id);
        // Nome do jogador
        jogad[tamJog].setNome(jogadores[1]);
        // Altura do jogador
        int altura = Integer.parseInt(jogadores[2]);
        jogad[tamJog].setAltura(altura);
        // Peso do jogador
        int peso = Integer.parseInt(jogadores[3]);
        jogad[tamJog].setPeso(peso);
        // Universidade do jogador
        jogad[tamJog].setUniversidade(jogadores[4]);
        // Ano de nascimento do jogador
        int anoNascimento = Integer.parseInt(jogadores[5]);
        jogad[tamJog].setAnoNascimento(anoNascimento);
        // Cidade do  jogador
        jogad[tamJog].setCidadeNascimento(jogadores[6]);
        // Estado do jogador
        jogad[tamJog].setEstadoNascimento(jogadores[7]);
    }

    public static void main(String[] args) throws Exception {

        //tempo inicial do código
        //long tempoInicial = System.currentTimeMillis();
        //----------------------------------------------//
        
        // PRIMEIRA PARTE DA LEITURA
        HashComRehash tabelaRehash = new HashComRehash();
        String IdsJogadores = "";
        BufferedReader entrada = new BufferedReader(new InputStreamReader(System.in));
        IdsJogadores = entrada.readLine();

        while(IdsJogadores.equals("FIM") != true){
            jogad[tamJog] = new Jogador();
            TratarString(ler(IdsJogadores));
            tamJog++;
            IdsJogadores = entrada.readLine();
        }

        for(int i =0 ; i< tamJog ; i++){
            tabelaRehash.inserir(jogad[i]);
        }
        
        // SEGUNDA PARTE DA LEITURA
        String NomeJogadores = "";
        int alturajog = 0;
        boolean saida = false;
        NomeJogadores = entrada.readLine();

        for(int j =0 ; j< tamJog ; j++){
            if(NomeJogadores.compareTo(jogad[j].getNome()) == 0){
                alturajog = jogad[j].getAltura();
                j = tamJog;
            }else{
                alturajog = 0;
            }
        }

        while(NomeJogadores.equals("FIM") != true){
            saida = tabelaRehash.pesquisar(alturajog ,NomeJogadores);
            if(saida == false){
                System.out.println(NomeJogadores + " NAO");
            }else{
                System.out.println(NomeJogadores + " SIM"); 
            }

            NomeJogadores = entrada.readLine();
            
            for(int j =0 ; j< tamJog ; j++){
                if(NomeJogadores.compareTo(jogad[j].getNome()) == 0){
                    alturajog = jogad[j].getAltura();
                    j = tamJog;
                }
            }
        }

        //arquivo de Matricula sequencial 
        //long tempoFinal = System.currentTimeMillis();
        //Arq.openWrite("matrícula_hashRehash.txt");
        //Arq.println("695161" + "\t" + (tempoFinal - tempoInicial) + "\t");
        //Arq.close();
    } 
}
